signalfd.c

```diff
@@ -0,0 +1,170 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "event2/event-config.h"
+
+#include <unistd.h>
+#include <sys/signalfd.h>
+
+#include "event2/event.h"
+#include "event-internal.h"
+#include "evmap-internal.h"
+#include "evsignal-internal.h"
+#include "evthread-internal.h"
+
+static int sigfd_add(struct event_base *, evutil_socket_t, short, short, void *);
+static int sigfd_del(struct event_base *, evutil_socket_t, short, short, void *);
+
+static const struct eventop sigfdops = {
+	"signalfd_signal",
+	NULL,
+	sigfd_add,
+	sigfd_del,
+	NULL,
+	NULL,
+	0, 0, 0
+};
+
+static void
+sigfd_cb(evutil_socket_t fd, short what, void *arg)
+{
+	struct signalfd_siginfo fdsi;
+	struct event_base *base = arg;
+	ssize_t ret = read(fd, &fdsi, sizeof(fdsi));
+
+	EVUTIL_ASSERT(ret == sizeof(fdsi));
+	EVUTIL_ASSERT(fdsi.ssi_signo > 0 && fdsi.ssi_signo < NSIG);
+	EVUTIL_ASSERT(base->sig.ev_sigevent[fdsi.ssi_signo] != NULL);
+
+	EVBASE_ACQUIRE_LOCK(base, th_base_lock);
+	evmap_signal_active_(base, fdsi.ssi_signo, 1);
+	EVBASE_RELEASE_LOCK(base, th_base_lock);
+}
+
+static void
+sigfd_free_sigevent(struct event_base *base, int signo)
+{
+	int ret;
+	struct event* sigev = base->sig.ev_sigevent[signo];
+
+	EVUTIL_ASSERT(sigev != NULL);
+	/* Not sure about blocking here. */
+	event_del_nolock_(sigev, EVENT_DEL_AUTOBLOCK);
+	ret = close(sigev->ev_fd);
+	EVUTIL_ASSERT(ret == 0);
+	mm_free(sigev);
+	base->sig.ev_sigevent[signo] = NULL;
+}
+
+static int
+sigfd_add(struct event_base *base, int signo, short old, short events, void *p)
+{
+	int sigfd;
+	sigset_t mask;
+	struct event* sigev;
+
+	/* EV_SIGNAL event passed from evmap_signal_add() when setting
+           up and from evmap_signal_reinit_iter_fn() during reinit. */
+	EVUTIL_ASSERT(p != NULL);
+
+	EVUTIL_ASSERT(signo > 0 && signo < NSIG);
+	sigev = base->sig.ev_sigevent[signo];
+
+	if (sigev != NULL) {
+		if (old) {
+			/* We're performing reinit after fork(). */
+			sigfd_free_sigevent(base, signo);
+		} else {
+			/* We have an active signal fd
+			   for this signal already. */
+			return 0;
+		}
+	}
+
+	sigemptyset(&mask);
+	sigaddset(&mask, signo);
+	if (sigprocmask(SIG_BLOCK, &mask, NULL)) {
+		event_warn("sigprocmask() failed");
+		return -1;
+	}
+
+	sigfd = signalfd(-1, &mask, SFD_NONBLOCK);
```

Comment:
I wish we add `SFD_CLOEXEC` here, but child inherit signal disposition that was created with `signal`/`sigaction`. Let's put something like this in the comment.

Reply:
Well, using `fork()` doesn't imply the following `exec()`, so why not always using `SFD_CLOEXEC` here, much like `EPOLL_CLOEXEC` is used in the epoll backend?